### PR TITLE
detectors/gcp: support Cloud Run jobs

### PR DESCRIPTION
**Pull request by zchee**
Support Cloud Run jobs using `CLOUD_RUN_JOB` and `CLOUD_RUN_EXECUTION`.

### CODE REVIEW

**Review by damemi — COMMENTED** [review]
/gcbrun

**Comment by zchee** [comment]
Note that Cloud Run jobs do not add `K_XXX` environment variables automatically. Instead of `CLOUD_RUN_JOB` and `CLOUD_RUN_EXECUTION`.

**Comment by zchee** [comment]
@damemi not running gcbrun...?

**Comment by damemi** [comment]
hm that seemed to work

**Comment by zchee** [comment]
thanks

**Comment by damemi** [comment]
cc @dashpole

**Comment by dashpole** [comment]
I don't see those set in https://cloud.google.com/run/docs/container-contract#jobs-env-vars.  Are the docs incorrect?

**Comment by zchee** [comment]
@dashpole I think doc is lacking. I was tested on real environment.

File: detectors/gcp/detector.go
Comment (dashpole) [inline detectors/gcp/detector.go:39]: I don't think we need a new cloud run jobs constant here, since the detector will use the same functions (e.g. FaasName) to add resource attributes.

Comment (dashpole) [reply detectors/gcp/detector.go:39]: Can you instead update the onCloudRun to return true on cloudRunJobs as well?

**Comment by zchee** [comment]
@dashpole PTAL

**Comment by zchee** [comment]
@dashpole gentle ping

**Comment by dashpole** [comment]
Yes, sorry.  We are discussing as a team to make sure this is the right approach.

**Comment by zchee** [comment]
@dashpole Ah, I see. I waiting your discuss result.

**Comment by dashpole** [comment]
Hey @zchee, if you make a job with multiple tasks, what is the value in CLOUD_RUN_EXECUTION?  Does each task end up with the same job name and execution?  If so, we may need some additional information to prevent collisions between executions.

**Comment by zchee** [comment]
@dashpole I'll investigate it.

Just in case, in the case of the `gcloud bate run jobs create` command, it means `--tasks=` flag, not `--parallelism` flag, right?

**Comment by dashpole** [comment]
If you can, try both, and also look at what CLOUD_RUN_TASK_INDEX and CLOUD_RUN_TASK_ATTEMPT are set to.  Really appreciate the help

**Comment by zchee** [comment]
@dashpole okay, will check both situations today.

Also, I'll report Cloud Run jobs docs lacking that's 2 envs via Mercari's Google TAM (I work for Mercari, Inc).

**Comment by zchee** [comment]
@dashpole replied from Google support, said "seems to serverless team forgot update document", So I think would be better to waiting to update. Maybe your concern is also cleared.

**Comment by zchee** [comment]
@dashpole I found otel side hardcoded `K_SERVICE`, need fix it also.

- https://github.com/open-telemetry/opentelemetry-go-contrib/blob/v1.9.0/detectors/gcp/cloud-run.go#L110

**Comment by dashpole** [comment]
Note that the old cloud run (which [that code](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/v1.9.0/detectors/gcp/cloud-run.go#L110) references) is deprecated in favor of this code base.

**Comment by zchee** [comment]
@dashpole updated document

**Comment by zchee** [comment]
@dashpole gentle ping

I want to hear your opinions.

**Comment by dashpole** [comment]
What were CLOUD_RUN_TASK_INDEX and CLOUD_RUN_TASK_ATTEMPT set to?

Also, what is in d.metadata.InstanceID() (faas.id)?

Basically, we want to make sure that faas.name, faas.version, and faas.id uniquely identify an instance of a cloud run job.  Otherwise, if two things are running in parallel with the same faas.* attributes, there could be conflicts.

**Comment by zchee** [comment]
@dashpole I'll reply details later.

**Comment by zchee** [comment]
@dashpole 

Use https://github.com/GoogleCloudPlatform/golang-samples/tree/main/run/jobs

```diff
diff --git a/run/jobs/main.go b/run/jobs/main.go
index 05e3195..a8a6d0a 100644
--- a/run/jobs/main.go
+++ b/run/jobs/main.go
@@ -21,6 +21,8 @@ import (
 	"os"
 	"strconv"
 	"time"
+
+	"cloud.google.com/go/compute/metadata"
 )
 
 type Config struct {
@@ -99,6 +101,12 @@ func main() {
 	}
 
 	log.Printf("Completed Task #%s, Attempt #%s", config.taskNum, config.attemptNum)
+
+	instanceID, err := metadata.InstanceID()
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Printf("instanceID: %s\n", instanceID)
 }
 
 // Throw an error based on fail rate
```

```console
$ gcloud alpha run jobs create --tasks=5 --parallelism=3 --region=asia-northeast1 jobs
# execute...
```

```
labels: {
run.googleapis.com/execution_name: "jobs-ph5p9"
run.googleapis.com/task_attempt: "0"
run.googleapis.com/task_index: "0"
}
```

```
labels: {
run.googleapis.com/execution_name: "jobs-ph5p9"
run.googleapis.com/task_attempt: "0"
run.googleapis.com/task_index: "1"
}
```

```
labels: {
run.googleapis.com/execution_name: "jobs-ph5p9"
run.googleapis.com/task_attempt: "0"
run.googleapis.com/task_index: "2"
}
```

```
textPayload: "2022/09/10 08:28:37 instanceID: 00c527f6d4ab9f193b199268ccb9d99d05ebc33d8983a56d2faef9169e9b42adf8fb6caaf62a35d1bff6df6b375c4282310b634da8218f77e706845ac6f4d5f27714"
```

**Comment by dashpole** [comment]
So, it looks like we have:

Just job name
`CLOUD_RUN_JOB=my-job-name`

job name with execution suffix
`CLOUD_RUN_EXECUTION=my-job-name-ph5p9`

task index
`CLOUD_RUN_TASK_INDEX=0`

job UID
`instanceID=00c527f6d4ab9f193b199268ccb9d99d05ebc33d8983a56d2faef9169e9b42adf8fb6caaf62a35d1bff6df6b375c4282310b634da8218f77e706845ac6f4d5f27714`

Edit: just realized that they updated the docs to show what each environment variable has: https://cloud.google.com/run/docs/container-contract#jobs-env-vars.

**Comment by dashpole** [comment]
My current thought is that we don't really have anything that could be construed as `faas.version`, right?  Seems like we only have `faas.id`, and `faas.name`, but no `faas.version`...

**Comment by dashpole** [comment]
Looking at this some more, it does look like the execution is basically the version, in that it change each time a new version is rolled out.

It would be really nice to include the CLOUD_RUN_TASK_INDEX somewhere, as then you could use execution + index to uniquely identify a job

**Comment by zchee** [comment]
@dashpole okay, fix asap

**Comment by zchee** [comment]
@dashpole use `fmt.Sprintf("%s/%s", version, index)`, PTAL.

**Comment by dashpole** [comment]
I discussed this with the team, and unfortunately I think we will need to update the OTel spec for cloud run jobs before we can merge this.  https://github.com/open-telemetry/opentelemetry-specification/blob/94c9c75c4f82fbda08bddff02ce9a0c582fdd55c/specification/resource/semantic_conventions/faas.md

**Comment by zchee** [comment]
@dashpole 
> I discussed this with the team, and unfortunately I think we will need to update the OTel spec for cloud run jobs before we can merge this

I see.
Who is contribute this change to otel?

**Comment by zchee** [comment]
@dashpole gentle ping

**Comment by damemi** [comment]
Hi @zchee, sorry this got lost. @dashpole is on leave at the moment so I'll pick this back up.

Catching up, I don't think there were any issues with this besides adding the convention for Cloud Run Job versions to the specification. I've opened a PR there to do this: https://github.com/open-telemetry/opentelemetry-specification/pull/3378

Once we merge that, I think we can move forward with this again. Thanks for your patience

**Comment by zchee** [comment]
@damemi Oh wow, that's great news :D And thanks for creating PR to OTel side.
I'll waiting for merge above with resolve conflicts.

**Comment by zchee** [comment]
@damemi fixed lint issue, PTAL.

**Comment by damemi** [comment]
/gcbrun
thanks @zchee! just going to wait for https://github.com/open-telemetry/opentelemetry-specification/pull/3378, but I'm tracking this now so it won't get lost again :)

File: detectors/gcp/detector.go
Comment (zchee) [inline detectors/gcp/detector.go:51]: just sorted by enum order.

**Comment by zchee** [comment]
@damemi Failed require `ops-go-e2e-cloud-functions` integration test. just flaky, or should I fix something?

File: detectors/gcp/detector.go
Comment (damemi) [inline detectors/gcp/detector.go:52]: Ah I think this needs to have cloud functions first, see https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/pull/552:

> The detector was incorrectly detecting the cloud run platform instead of the cloud functions platform. This is because cloud functions v2 is implemented on top of cloud run. This commit makes the detector check for cloud functions first to fix this issue.

We should at least leave a comment here about why this ordering is important. It would be better if we can find a way to more soundly fix this rather than relying on the order, but I'm not sure that will be possible if cloud functions is built on cloud run.

**Review by damemi — CHANGES_REQUESTED** [review]

File: detectors/gcp/detector.go
Comment (zchee) [reply detectors/gcp/detector.go:52]: I remembered. will fix

**Comment by damemi** [comment]
@zchee sorry for the delay in this. The spec changes just merged and from the discussion in https://github.com/open-telemetry/opentelemetry-specification/pull/3378 this will be slightly different than we talked about here.

Instead, we will have separate GCP-specific resource attributes for Cloud Run execution and Job ID (https://github.com/open-telemetry/opentelemetry-go/pull/4079 will add these to the Go semconv package).

I think we should probably just make these available via their own functions, and maybe add 2 new functions `IsJob()` and `IsService()`. wdyt?

**Comment by zchee** [comment]
make sense. I'll implements.

File: detectors/gcp/detector.go
Comment (zchee) [reply detectors/gcp/detector.go]: Cloud Functions gen2 also reserved the `K_CONFIGURATION` environment variable, but undocumented... :(
Well, therefore wrong detection if change the `switch...case` order.

As you said, we should leave a details comment here.

**Review by damemi — CHANGES_REQUESTED** [review]
Thanks for updating this @zchee. We just need to make a few changes to match [the spec for Cloud Run Jobs](https://github.com/open-telemetry/semantic-conventions/blob/8fc3d034c7bdf4c975fa64a436ea92ef5192cccc/specification/resource/semantic_conventions/cloud_provider/gcp/cloud_run.md) that was approved upstream.

They did not like the idea of using `execution+task` as `faas.version`, so those will be their own individual attributes instead. That means we need separate functions for each of them.

To know that we need to call those functions, Cloud Run Jobs have to be treated like their own environment, separate from Services. So a new detector function for that as well.

imo, `faas.name` is fine to be shared between Jobs and Services. These are semantically providing the same info, even though it is technically different environment variables.

To summarize, we need the following new functions:
* `onCloudRunJob()` - [detect](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/9c6bfe4427dee294b17f8311c521f24072c26bf3/detectors/gcp/detector.go#L51) if we are running on Cloud Run Jobs
* `FaasJobExecution()` - return the Execution ID of the Job
* `FaasJobTaskIndex()` - return the Task Index for the execution

Our [Collector resource detection processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/0f8e1282b5cc7b8c1639d3117db36a5d31a3cccd/processor/resourcedetectionprocessor/internal/gcp/gcp.go#L61) will be updated to write to the new `gcp.cloud_run.*` attributes when running on Cloud Run Jobs, and our exporter's [monitored resource mapping](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/9c6bfe4427dee294b17f8311c521f24072c26bf3/internal/resourcemapping/resourcemapping.go#L135) will be updated to parse `execution+task` into a unique task ID.

**Comment by zchee** [comment]
> We should at least leave a comment here about ...

https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/pull/465#discussion_r1225734878

Should I include that to this PR?

**Comment by damemi** [comment]
> > We should at least leave a comment here about ...
> 
> [#465 (comment)](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/pull/465#discussion_r1225734878)
> 
> Should I include that to this PR?

yes please do, this will prevent the same headache in the future :)

for the naming inconsistencies, I don't really have a good answer. I think the usage depends on the context, ie when referring to the product "Cloud Run jobs" makes sense, when referring to multiple jobs the plural makes sense vs a singular job. Either way, I would always prefer the `cloudRunJob` camel case vs `cloudRunjob` in code

File: detectors/gcp/faas.go
Comment (damemi) [reply detectors/gcp/faas.go:75]: Something shorter like `JobExecution()` and `JobTaskIndex()` are probably fine too

Comment (zchee) [reply detectors/gcp/faas.go]: I choose the first one.

**Comment by zchee** [comment]
@damemi PTAL.

**Review by damemi — APPROVED** [review]
thanks @zchee for working on this!

**Pull request merged**
